src/Orleans.Core/Timers/TimerManager.cs

```diff
@@ -0,0 +1,363 @@
+    {
+        public static Task Delay(TimeSpan timeout) => DelayUntil(DateTime.UtcNow + timeout);
+
+        public static Task DelayUntil(DateTime dueTime)
+        {
+            var result = new DelayTimer(dueTime);
+            TimerManager<DelayTimer>.Register(result);
+            return result.Completion;
+        }
+
+        private sealed class DelayTimer : ITimerCallback, ILinkedListElement<DelayTimer>
+        {
+            private readonly TaskCompletionSource<int> completion =
+                new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            public DelayTimer(DateTime dueTime)
+            {
+                this.DueTime = dueTime;
+            }
+
+            public Task Completion => this.completion.Task;
+
+            public DateTime DueTime { get; }
+
+            public bool IsCanceled => false;
+
+            public void OnTimeout() => this.completion.TrySetResult(0);
+
+            DelayTimer ILinkedListElement<DelayTimer>.Next { get; set; }
+        }
+    }
+
+    /// <summary>
+    /// Manages timers of a specified type, firing them after they expire.
+    /// </summary>
+    /// <typeparam name="T">The timer type.</typeparam>
+    internal static class TimerManager<T> where T : class, ITimerCallback, ILinkedListElement<T>
+    {
+        /// <summary>
+        /// The maximum number of times a queue can be denied servicing before servicing is mandatory.
+        /// </summary>
+        private const int MaxStarvation = 2;
+
+        /// <summary>
+        /// The number of milliseconds between timer servicing ticks.
+        /// </summary>
+        private const int TimerTickMilliseconds = 50;
+
+        /// <summary>
+        /// Lock protecting <see cref="allQueues"/>.
+        /// </summary>
+        // ReSharper disable once StaticMemberInGenericType
+        private static readonly object AllQueuesLock = new object();
+
+        // ReSharper disable once StaticMemberInGenericType
+        // ReSharper disable once NotAccessedField.Local
+        private static readonly Timer QueueChecker;
+
+        /// <summary>
+        /// Collection of all thread-local timer queues.
+        /// </summary>
+        private static ThreadLocalQueue[] allQueues = new ThreadLocalQueue[16];
+
+        /// <summary>
+        /// The queue for the current thread.
+        /// </summary>
+        [ThreadStatic]
+        private static ThreadLocalQueue threadLocalQueue;
+
+        static TimerManager()
+        {
+            var timerPeriod = TimeSpan.FromMilliseconds(TimerTickMilliseconds);
+            QueueChecker = new Timer(_ => CheckQueues(), null, timerPeriod, timerPeriod);
+        }
+
+        /// <summary>
+        /// Registers a timer.
+        /// </summary>
+        public static void Register(T timer)
+        {
+            ExpiredTimers expired = null;
+            var queue = EnsureCurrentThreadHasQueue();
+            queue.Lock.Get();
```

Comment:
Shouldn't this line be inside the `try` to ensure `queue.Lock.Release()` will always execute?

Reply:
Perhaps I should add a `TryRelease` which only releases if the current thread actually holds the lock?

Reply:
My concern is the opposite - if the `try` doesn't execute (for whatever reason) the lock would never be released. I think putting the `Lock.Get()` line inside the try *should* guarantee that the lock will always get released.

Reply:
That depends on where the assumed exception comes from. If it came from `SpinWait` then putting `Get` into the lock and *not* having try-release semantics could be incorrect.

Reply:
So I did both anyhow. The return value of `TryRelease` is never used and it will always succeed if the lock was acquired.